clio-server/src/main/scala/org/broadinstitute/clio/server/service/PersistenceService.scala

```diff
@@ -51,10 +55,59 @@ class PersistenceService private (persistenceDAO: PersistenceDAO,
       document.upsertId
     }
   }
+
+  private val countZero = Future.successful(0L)
+
+  /**
+    * Recover missing documents for a given Elasticsearch index
+    * from the source of truth.
+    *
+    * Returns the number of updates pulled & applied on success.
+    *
+    * Fails immediately if an update pulled from the source of
+    * truth fails to be applied to the search index.
+    */
+  def recoverMetadata[D <: ClioDocument: Indexable: HitReader: Decoder](
+    index: ElasticsearchIndex[D]
+  )(implicit ec: ExecutionContext): Future[Long] = {
+    searchDAO.getMostRecentDocument(index).flatMap { maybeLastDocument =>
+      maybeLastDocument.fold(countZero) { lastDocument =>
+        for {
+          documents <- persistenceDAO.getAllSince(lastDocument.upsertId, index)
+          count <- updateAndCount(documents, index)
+        } yield {
+          count
+        }
+      }
+    }
+  }
+
+  /**
+    * Apply a sequence of partial documents to an Elasticsearch index,
+    * returning the total number of documents applied.
+    *
+    * Updates run serially, and short-circuit if applying any one
+    * update fails.
+    */
+  private def updateAndCount[D <: ClioDocument: Indexable](
+    documents: Seq[D],
+    index: ElasticsearchIndex[D]
+  )(implicit ec: ExecutionContext): Future[Long] = {
+    documents.foldLeft(countZero) {
+      case (countFut, document) => {
+        for {
+          count <- countFut
+          _ <- searchDAO.updateMetadata(document, index)
+        } yield {
+          count + 1
+        }
+      }
+    }
+  }
 }
 
 object PersistenceService {
-  def apply(app: ClioApp): PersistenceService = {
+  def apply(app: ClioApp)(implicit mat: Materializer): PersistenceService = {
```

Comment:
Is it obvious to most readers why this implicit is here? It isn't obvious to me.

Reply:
It's implicit because the `PersistenceService` constructor needs it to be implicit, because all of the Akka stream-y things need it to be implicit. Usually when I see an implicit on an `apply` method with no apparent use I assume it's dependency injection.

Reply:
Yeah, my problem is that if I don't know what needs it, I won't know when I can remove it. But I guess you get an error if it's needed, so you'll know when you need to add it? I guess it would be nice to get a warning in the case where a declared implicit is never used.

Reply:
The compiler will yell on an unused implicit parameter just like it does for unused explicit parameters.